zkevm-circuits/src/util/word.rs

```diff
@@ -231,6 +364,11 @@ impl<F: Field> Word<Expression<F>> {
     pub fn sub_unchecked(self, rhs: Self) -> Self {
         Word::new([self.lo() - rhs.lo(), self.hi() - rhs.hi()])
     }
+
+    /// No overflow check
+    pub fn expr_unchecked(&self) -> Expression<F> {
+        self.lo() + self.hi() * (1 << (N_BYTES_HALF_WORD * 8)).expr()
+    }
```

Comment:
There are many, can see one example in evm circuit word lo/hi refactor [TransferGadget](`https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1411/files#diff-e72a6a818cbda6c14cb551573432e59fe61ce109345dd4c088508a60c27979afL178-L183`) Just be patient 😄 It took 10 secs on my browser to nativage the the cursor haha

Basically this is to handle `address`, such as caller_address, callee_address, coinbase. We have many gadgets take address as single expression. This function is to use to convert address from word to single expression. And who use/review this function to be aware.  

Our principle is, if there is `Word<Expression<F>>` type and need to put address, then it should be 16 bytes on `lo` and 4 bytes on `hi`.  But if there just require a single expression, we can convert address to single expression with this function.

Reply:
Thanks for pointing out the link.
I checked out the use cases and now see the rationale for having this function.

However, I still think the use `expr_unchecked` doesn't make sense most of the time, or at least we shouldn't use them directly in the circuit. There are some reasons:

With the word lo-hi refactor, we can represent addresses as words in as many places as possible to minimize the effort for conversion. It would be something similar In spec PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/424/.

One of the reasons we need to use `expr_unchecked` to convert an address from a word is that many of our API returns address as a word. For example, we got `callee_address` as a word because the API `call_context_read_as_word` now returns a word.

As address as a word output become common, it would be convenient for the API to take an address as a word as input. The API knows that the input should be an address, so doing the  `expr_unchecked` conversion inside that API is safe. Pushing the conversion decision to circuit developers is more error-prone.

I checked the use cases in the evm circuit PR, the conversions happen in gadgets like UpdateBalanceGadget and TransferGadget and APIs like `account_storage_read` or `account_read`. They all propagate the address all the way to RWtable lookup. So one best destination for the conversion might be at the rw_lookup API.

If we look at the [RWTable](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#rw_table), the only place we can consider compressing the address from a word to a field element is "4 Address k2", which is this place https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/review-commments-word-lo-hi-docs-util/zkevm-circuits/src/evm_circuit/table.rs#L332 in the circuit.

To summarize, I propose the following:

- we change the function signatures of whatever API takes addresses as inputs to take them as a word.
- we compress the address at the level of  `fn input_exprs(&self)` of the `Lookup` struct or somewhere that low level.

Reply:
Pretty agree with your proposal! For split address in word hi/lo, why we need to prioritize it is because it's more like a bug since in word expression we do not expect to have lo > 128 bits. Here is more like a consistent address representation across the codebase.

How about 
- mark this function as `deprecated` for the unsafe reason for now.
- then probably removed it in the following evm_circuit pr along with rw table fix ~?